### PR TITLE
Reimplement channel_listen as context manager

### PR DESCRIPTION
**Pull request by moritz89**
Why:

- Allow user code to be run after Channels subscription
- Avoids race condition when confirming GQL subscriptions

This change addresses the need by:

- Returning the async generator instead of runnning it

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/strawberry-graphql/strawberry/issues/2799

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by moritz89** [comment]
Addresses issue #2799

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @moritz89 for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by moritz89** [comment]
This is the simplest implementation possibility. However, it has the problem that a leak can occur if the async generator is never started. Will create a second approach that uses a context manager to ensure that the cleanup will be done.

**Comment by moritz89** [comment]
Added a second option of starting channel_listen with a context manager to ensure that group_discard is called after it is started.

**Comment by moritz89** [comment]
@DoctorJohn @patrick91 Wanted to ask which approach you'd prefer? ContextManager or the simple approach

**Comment by DoctorJohn** [comment]
> @DoctorJohn @patrick91 Wanted to ask which approach you'd prefer? ContextManager or the simple approach

Hi @moritz89 thanks for the PR. I prefer the ContextManager approach. It's the one I did my quick proof of concept with. Main argument for it is the cleanup ability you already mentionened in your comment. Feel free to remove the "simple approach".

**Comment by moritz89** [comment]
Isn't this fun, adding breaking changes :sweat_smile: Anything else that should be updated?

**Comment by patrick91** [comment]
@moritz89 it is a lot of work to make it backwards compatibile with a deprecation? 😊

**Comment by moritz89** [comment]
> @moritz89 it is a lot of work to make it backwards compatibile with a deprecation? blush

Not at all. The question is only how this should be handled. It would result in two functions in the API and the question is how they should be named. One option is to append `_ctx` to the new function.

**Comment by moritz89** [comment]
@patrick91 I've had a look at unifying the calls in the following style, but also that did not work and going down that line of thought makes the library code itself more and more tricky. If you have any suggestions, would be glad to take them up.

```
class ChannelListen:
    def __init__(
        self, type: str, *, timeout: Optional[float] = None, groups: Sequence[str] = ()
    ) -> None:
        ...

    async def __call__(
        self, type: str, *, timeout: Optional[float] = None, groups: Sequence[str] = ()
    ) -> Any:
        ...

    async def __aenter__(self) -> Awaitable[AsyncGenerator[Any, None]]:
        ...

    async def __aexit__(self) -> None:
        ...

    async def _channel_listen_generator(
        self, queue: asyncio.Queue, timeout: Optional[float]
    ) -> AsyncGenerator[Any, None]:
        ...

class ChannelsConsumer(AsyncConsumer):
    ...
    channel_listen = ChannelListen
```

**Comment by moritz89** [comment]
Fixed mypy issues and updated existing tests and added a new one, but don't know why one is failing:

`ERROR 😡  tests/cli/test_server.py::test_debug_server_routes - KeyError: 'STRAWBERRY_DEBUG_SERVER_SCHEMA'`

Would appreciate help in fixing it

**Comment by moritz89** [comment]
How should the deprecation be communicated to the library users?

**Comment by DoctorJohn** [comment]
> I'll create a commit assuming you meant the former (deprecate the existing approach).

Yup, that's what I meant. I think the new approach is superior and covers all usecases of the old one.

> How should the deprecation be communicated to the library users?

Using the `warnings` module, just like here for example:

https://github.com/strawberry-graphql/strawberry/blob/19be5751c888bea76c03401be105a4a1b80c723c/strawberry/types/info.py#L52-L56

**Comment by moritz89** [comment]
@patrick91 The MR won't break compatibility but adds a deprecation warning

File: docs/integrations/channels.md [hunk 1]
```diff
@@ -136,13 +136,15 @@ class Subscription:
...
-                )
+        async with ws.channel_listen("chat.message", groups=room_ids) as cm:
+            # Optionally: Confirm subscription with `yield None`
+            async for message in cm:
+                if message["room_id"] in room_ids:
+                    yield ChatRoomMessage(
+                        room_name=message["room_id"],
+                        message=message["message"],
+                        current_user=user,
+                    )
```
Comment (DoctorJohn) [inline docs/integrations/channels.md:147]: These docs need an update now :)

```suggestion
        async with ws.listen_to_channel("chat.message", groups=room_ids) as cm:
            async for message in cm:
                if message["room_id"] in room_ids:
                    yield ChatRoomMessage(
                        room_name=message["room_id"],
                        message=message["message"],
                        current_user=user,
                    )
```

File: docs/integrations/channels.md [hunk 2]
```diff
@@ -153,7 +155,8 @@ message to all the channel_layer groups (specified in the subscription argument
 <Note>
 
 The `ChannelsConsumer` instance is shared between all subscriptions created in
-a single websocket connection. The `ws.channel_listen` function will yield all
+a single websocket connection. The `ws.channel_listen` context manager will return
```
Comment (DoctorJohn) [inline docs/integrations/channels.md:158]:
```suggestion
a single websocket connection. The `ws.listen_to_channel` context manager will return
```

Here as well :)

**Review by DoctorJohn — CHANGES_REQUESTED** [review]

**Comment by patrick91** [comment]
> @patrick91 The MR won't break compatibility but adds a deprecation warning

Thank you so much! <3

**Comment by moritz89** [comment]
@DoctorJohn @patrick91 Does this MR require anything else or can it be merged?

File: strawberry/channels/handlers/base.py [hunk 3]
```diff
@@ -144,6 +146,79 @@ async def channel_listen(
...
+                at the end of the execution.
+        """
+
+        # Code to acquire resource (Channels subscriptions)
+        if self.channel_layer is None:
+            raise RuntimeError(
+                "Layers integration is required listening for channels.\n"
+                "Check https://channels.readthedocs.io/en/stable/topics/channel_layers.html "  # noqa:E501
+                "for more information"
+            )
```
Comment (patrick91) [inline strawberry/channels/handlers/base.py:181]: 👍

File: RELEASE.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+class MyDataType:
+   name: str
+
+@strawberry.type
+class Subscription:
+   @strawberry.subscription
+   async def my_data_subscription(
+      self, info: Info, groups: list[str]
+   ) -> AsyncGenerator[MyDataType | None, None]:
+      yield None
```
[the remark this answers is not in the document]
Comment (moritz89) [reply RELEASE.md:26]: It's to show how GQL subscription confirmations (with null data) should be migrated, i.e., that confirmation messages should be sent before the `async for` loop. However, for every else not confirming their GQL subscriptions the `yield None` can be omitted in both new and old code variants.

How should the doc be updated to reflect that?

[the remark this answers is not in the document]
Comment (moritz89) [reply RELEASE.md:26]: Good point. Added a section in the channels.md doc

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**